packages/flutter/lib/src/foundation/tree_diagnostics_mixin.dart

```diff
@@ -2,9 +2,1535 @@
+
+    String result = valueToString();
+    if (result.isEmpty && emptyDescription != null)
+      result = emptyDescription;
+    return addTooltip(result);
+  }
+
+  String addTooltip(String text) {
+    return tooltip == null ? text : '$text ($tooltip)';
+  }
+
+  /// Description if the property [value] is null.
+  final String nullDescription;
+
+  /// Description if the property description would otherwise be empty.
+  final String emptyDescription;
+
+  /// Optional tooltip typically describing the property.
+  ///
+  /// Example tooltip: 'physical pixels per logical pixel'
+  /// If present, the tooltip is added in parenthesis after the raw value when
+  /// generating the string description.
+  final String tooltip;
+
+  Type get propertyType => T;
+
+  /// Returns the value of the property either from cache or by invoking a
+  /// [ComputePropertyValueCallback].
+  ///
+  /// If an exception is thrown invoking the [ComputePropertyValueCallback],
+  /// [value] returns `null` and the exception thrown can be found via the
+  /// [exception] property.
+  ///
+  /// See also:
+  ///
+  ///  * [valueToString], which converts the property value to a string.
+  T get value {
+    _maybeCacheValue();
+    return _value;
+  }
+
+  T _value;
+
+  bool _valueComputed;
+
+  @override
+  T get object => value;
+
+  Object _exception;
+
+  /// Exception thrown if accessing the property [value] threw an exception.
+  ///
+  /// Returns null if computing the property value did not throw an exception.
+  Object get exception {
+    _maybeCacheValue();
+    return _exception;
+  }
+
+  void _maybeCacheValue() {
+    if (_valueComputed)
+      return;
+
+    _valueComputed = true;
+    assert(_computeValue != null);
+    try {
+      _value = _computeValue();
+    } catch (exception) {
+      _exception = exception;
+      _value = null;
+    }
+  }
+
+  /// Optional defaultValue for the object.
+  ///
+  /// If [object] `==` [defaultValue] and [defaultValue] is not null then
+  /// the node is [hidden].
+  final Object defaultValue;
+
+  final bool _hidden;
+
+  @override
+  bool get hidden {
+    if (_hidden)
+      return true;
+
+    if (defaultValue != kNoDefaultValue) {
+      if (exception != null)
+        return false;
+
+      return object == defaultValue;
+    }
+    return false;
+  }
+
+  final ComputePropertyValueCallback<T> _computeValue;
+
+  @override
+  List<DiagnosticsNode> getProperties() => <DiagnosticsNode>[];
+
+  @override
+  List<DiagnosticsNode> getChildren() => <DiagnosticsNode>[];
+}
+
+/// Add additional properties describing an object.
+///
+/// See also:
+///
+///  * [TreeDiagnosticsMixin.debugFillProperties], which lists best practices
+///    for specifying properties.
+typedef void FillPropertiesCallback(List<DiagnosticsNode> properties);
+
+/// Returns a list of [DiagnosticsNode] objects describing an object's children.
+///
+/// See also:
+///  * [TreeDiagnosticsMixin.debugDescribeChildren], which lists best practices
```

Comment:
need a blank line between paragraphs and bulleted list for the docs to format correctly